Explainer for text conversion stuff (#1409)
WICG/turtledove: FLEDGE.md

```

### 8. Common Utilities

#### 8.1 String and Uint8Array conversion

> Any worklet script can convert from a String to a Uint8Array by calling

Reviewer: ```suggestion
Any bidding, scoring or reporting script can convert from a `String` to a `Uint8Array` by calling
```

Author: Done

> `protectedAudience.encodeUtf8(someString)`, which will return a UInt8Array containing the contents

Reviewer: ```suggestion
`protectedAudience.encodeUtf8(someString)`, which will return a `UInt8Array` containing the contents
```

Author: Done

of the string encoded as UTF-8.

> Similarly, a Uint8Array containing UTF-8 data can be converted to a String by calling
> `protectedAudience.decodeUtf8(someArray)`. Note that this is specifically for Uint8Arrays, and

Reviewer: ```suggestion
Similarly, a `Uint8Array` containing UTF-8 data can be converted to a String by calling
`protectedAudience.decodeUtf8(someArray)`. Note that this is specifically for `Uint8Arrays`, and
```

Author: Done

will not handle other, similar, types.

> 

Reviewer: ```suggestion

These utility functions are useful for passing `String`s into and out of WebAssembly functions where `String`s are required to pass through the [`WebAssembly.memory ArrayBuffer`](https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/JavaScript_interface/Memory).
```

Author: Did only the first part since I don't want to tell people with more WebAssembly experience than me how to interface with it...

Reviewer: We should probably include links to wasm-bindgen and TextEncoder and TextDecoder.

Author: > I think my previous suggested text is reasonable. 

Memory array isn't the only way.
https://developer.mozilla.org/en-US/docs/WebAssembly/Guides/JavaScript_builtins

(And your polyfill example lacks feature detection --- and is too modern a JS dialect for me to check by eyeballing)

Reviewer: > Memory array isn't the only way.

I think I covered that by proposing "may rely on..."

JS builtins seem like they might allow calling built-in JS functions on JS Strings from WASM, but I don't think they allow WASM to efficiently access the characters of the string.
I'm not saying my proposed text is perfect, feel free to edit and improve it, but I think we need to provide more help to readers in using these new additions and I think we need to explain the motivation for these.  Explainers' central responsibility is to discuss motivation.

Author: It has a "charCode at index X" builtin, but I have no idea of how fast it is; and also how well it fits the tooling on the other side. 

Thanks for the context/motivation.

Maybe I could say something like:

Tools like wasm-bindgen frequently perform these conversions using TextEncoder and TextDecoder interfaces in order to pass Strings from JavaScript to WASM efficiently. Since these classes are not available in the bidder, seller, or reporting script environments, protectedAudience.encodeUtf8 and decodeUtf8 functions provide a way of efficiently polyfilling the minimum needed subset of their functionality. For example, a version incorporating feature detection:
```
TextEncoder = function() {}
TextEndoder.prototype = {}
if (globalThis.protectedAudience && protectedAudience.encodeUtf8) {
   TextEncoder.prototype.encode = protectedAudience.encodeUtf8;
} else {
   TextEncoder.prototype.encode = slowerJavaScriptEncodeImplementation;
}
```

Author: Pushed this, so PTAL if you can (though I dunno if CMA pre-review process stuff is doable in half a day).
